src/services/refactors/convertReactPureComponent.ts

```diff
@@ -0,0 +1,294 @@
+            if (namedBindings && isNamespaceImport(namedBindings)) { result.React = namedBindings.name.text; }
+            // import d from "mod" => name = d, namedBinding = undefined
+            else if (name) { result.React = name.text; }
+            // import d, * as ns from "mod" => name = d, namedBinding: NamespaceImport = { name: ns }
+            // ? No this case
+            // import { a, b as x } from "mod" => name = undefined, namedBinding: NamedImports = { elements: [{ name: a }, { name: x, propertyName: b}]}
+            if (namedBindings && isNamedImports(namedBindings)) {
+                namedBindings.elements.forEach(e => {
+                    if (e.propertyName) {
+                        const p: keyof typeof result = e.propertyName.text as any;
+                        if (names.indexOf(p) !== -1) { result[p] = e.name.text; }
+                    }
+                    else {
+                        const n: keyof typeof result = e.name.text as any;
+                        if (names.indexOf(n) !== -1) { result[n] = n; }
+                    }
+                });
+            }
+            // import d, { a, b as x } from "mod" => name = d, namedBinding: NamedImports = { elements: [{ name: a }, { name: x, propertyName: b}]}
+            // ? No this case
+        });
+        if (result.React) {
+            result.Component = result.Component || result.React + ".Component";
+            result.PureComponent = result.PureComponent || result.React + ".PureComponent";
+            result.SFC = result.SFC || result.React + ".SFC";
+            result.StatelessComponent = result.StatelessComponent || result.StatelessComponent + ".StatelessComponent";
+        }
+        else if (!result.Component && !result.PureComponent && !result.React && !result.SFC && !result.StatelessComponent) {
+            return undefined;
+        }
+        return { ...result, importClause: importClause! };
+    }
+
+    function isReactSFCDeclaration(node: Node, sourcefile: SourceFile): Component | undefined {
+        if (!isSFCLikeDeclaration(node)) { return; }
+        if (node.asteriskToken) { return; }
+        if (!node.body) { return; }
+
+        // TODO: Should also check the actual type insteadof only receive explicit typed
+        const typeNode = node.type;
+        if (!typeNode) { return; }
+        const binding = getReferenceToReact(sourcefile)!;
+        const type = removeEmpty(typeNode.getFullText(sourcefile));
+        if (type !== binding.SFC && type !== binding.Component) { return; }
+
+        let propsType: TypeNode | undefined;
+        if (isTypeReferenceNode(typeNode)) {
+            if (typeNode.typeArguments) { propsType = typeNode.typeArguments[0]; }
+        }
+
+        let render: Block;
+        if (isBlock(node.body)) {
+            render = node.body;
+        }
+        else if (isExpression(node.body!)) {
+            render = createBlock([createReturn(node.body)]);

```

Comment:
do not create a block untill you are applying the edit.

Reply:
I got a problem with manipulating the AST and get the textchange, do I do something wrong? Thanks

https://github.com/Microsoft/TypeScript/issues/24751